plan/plan.go

```diff
@@ -175,3 +175,18 @@ func shouldUpdateTTL(desired, current *endpoint.Endpoint) bool {
 	}
 	return desired.RecordTTL != current.RecordTTL
 }
+
+func filterRecordsForPlan(records []*endpoint.Endpoint) []*endpoint.Endpoint {
+	filtered := []*endpoint.Endpoint{}
+
+	for _, record := range records {
+		switch record.RecordType {
```

Comment:
Does this only happens for TXT records? If so it would make sense to rewrite the logic IMHO.

Reply:
Well, my intent here is to be explicit that only certain records "conflict." The truth is, the only record that "conflicts" with anything is CNAME. From [RFC 1034](https://tools.ietf.org/html/rfc1034#page-15) (emphasis mine:)

>The domain system provides such a feature using the canonical name
(CNAME) RR.  A CNAME RR identifies its owner name as an alias, and
specifies the corresponding canonical name in the RDATA section of the
RR.  **If a CNAME RR is present at a node, no other data should be
present; this ensures that the data for a canonical name and its aliases
cannot be different.**  This rule also insures that a cached CNAME can be
used without checking with an authoritative server for other RR types.

So, if you have a CNAME record, then you MUST NOT have ANY other kind of record.

This actually brings a pretty interesting conundrum: It means that ExternalDNS can only create a CNAME record if no other record exists with this name, but it can create A records if any other record exists.

ExternalDNS today only cares about CNAME, A, and TXT records, and it only cares about TXT records for the TXT registry. To fix this bug *properly*, I argue one would have to include every kind of record and consider it. And in fact, I think that would probably not be a bad idea. However, it would only allow you to output a better error message, probably, since you can't actually do much to fix the problem itself from ExternalDNS's PoV.

So I opted to go for a simple solution of explicitly stating which records the planner cares about, which I believe is CNAME and A records. Specifically, the planner never wants to create CNAME records together with A records anyways, so logically this makes sense. Likewise, the planner probably never wants to delete a TXT record that it had nothing to do with, so this probably makes sense too.

As a sidenote... doesn't this mean the TXT registry should never have a blank prefix, since it could never work with CNAME records? I imagine this is probably an issue for Amazon ELB users.

Reply:
I see your point and I agree. 

There's currently also an IPv6 PR (#576) open, so it is also planned to support AAAA records.
Additionally to that there's another PR open #579 which maybe make sense to merge first. 

We have a logic for CNAME's in AWS which will be handled as A or AAAA records. It will create ALIAS records pointing to ELB's and a TXT record as well.

Could you add some comment above the filterRecordsForPlan func? 

Otherwise LGTM